### DISCUSSION TITLE
how to call a js function in another js engine safely

### DISCUSSION THREAD

**Discussion by Zaitonn** [sebastienros/jint#1560] [Q&A] [answered]
now I need to call a function in another js engine in order to create more features for my application

here is my way but maybe there are some errors i think :(

```mermaid
flowchart TD
e1("engine1(JS1)")
e2("engine2(JS2)")
e3("engine3(JS3)")

e1-- export functions -->e2
e1-- export functions -->e3

c1["lock engine1 to call functions which are exported from it in c#"]
e2 --> c1
e3 --> c1
c2["return values to engine2/3"]

c1--->c2
```

- C#
    ```cs
    /// <summary>
    /// call the specified function with some arguments
    /// </summary>
    /// <returns>result</returns>
    public static JsValue SafeCall(JsValue func, params JsValue[] arguments)
    {
        if (func is not FunctionInstance functionInstance)
        {
            return JsValue.Undefined;
        }
        Engine engine = functionInstance.Engine;
        bool lockTaken = false;
        Monitor.TryEnter(engine, 1000, ref lockTaken);
        if (!lockTaken)
        {
            throw new TimeoutException();
        }
        try
        {
            return engine.Invoke(functionInstance, arguments);
        }
        finally
        {
            Monitor.Exit(engine);
        }
    }
    ```
- Engine1
    ```js
    function regCommand(handler) {
        // some checking and handling
        return false; // or true
        // in a word, this function will return a boolean value
    }

    serein.exports('CHregCommand', regCommand);

    // some indifferent codes are hidden XD
    ```
- Engine2
    ```js
    const CHregCommand = serein.imports('CHregCommand');

    const handler = {
        name: 'test1',
        triggers:[]
    }
    serein.log(serein.safeCall(CHregCommand, handler))
    ```
- Engine3
    ```js
    const CHregCommand = serein.imports('CHregCommand');

    const handler = {
        name: 'test2', // the name is different
        triggers:[]
    }
    serein.log(serein.safeCall(CHregCommand, handler))
    ```

`serein.exports` and `serein.imports` may not be important. they two only set/get value from a `Dictionary<string, JsValue>`

i expect that the two engine(2/3) should correctly call the function `CHregCommand` from engine 1 and get an output of boolean value  
but actually it seems that only one of the two can call it but without the output.

I have tried single-step debugging in C#, and it turned out that the code is stuck at line `return engine.Invoke(functionInstance, arguments);`. I guess `SafeCall` hasn't returned yet because the code in `finally` hasn't been executed. and this results that the another engine couldn't access to engine1 and then it threw `TimeoutException`

are there any mistakes? :(( i have been puzzled at it for days
or are there some alternative ways to reach the purpose?

---

if i remove `Monitor.TryEnter(engine, 1000, ref lockTaken);` and related checks, it can return correctly. 

---

jint: `3.0.0-beta-2048`
net: tested in net472 and net6.0-windows
os: windows 10

**Comment by Zaitonn** [comment]
https://github.com/sebastienros/jint/assets/103164490/7b811113-6e5d-4550-9ade-6365cdb84af8

**Reply by Genteure** [reply]
This file does not seems to be playable.

**Reply by Zaitonn** [reply]
try downloading?
i can play it on my phone
![E4A6DB6A-04BD-4F35-8E4F-CF9E0F7C291E](https://github.com/sebastienros/jint/assets/103164490/d2d89998-c52c-4e77-af76-a92b92aa11e5)


下载试试？（如果放不了我等下重新传一下

**Reply by Zaitonn** [reply]
https://github.com/sebastienros/jint/assets/103164490/f4ba7d9c-7c5f-40f2-a22a-17bb9a864163

**Comment by Genteure** [comment, marked as answer]
Would you benefit from having multiple instances of `Engine`? If single thread performance is acceptable, having only one `Engine` would greatly reduce the complexity.

If multiple instances are necessary, I would recommend writing the "bridge" between `Engine`s purely with C# primitive types and strings. Every "Jint object" holds a reference to their "owning" `Engine`, and creates other objects from that `Engine`. I could be wrong but I don't think mixing multiple `Engine`s together is a good idea or even a supported use case, you would have a bird nest mesh of objects coming from different `Engine`s mixed together.

Checkout [Worker API](https://developer.mozilla.org/en-US/docs/Web/API/Worker) for inspirations on implementing APIs that passes data between scripting contexts, specifically [`postMessage()` method](https://developer.mozilla.org/en-US/docs/Web/API/Worker/postMessage) and [`message` event](https://developer.mozilla.org/en-US/docs/Web/API/Worker/message_event). Another keyword to search for is "message bus" or "event bus".
You could then optionally write some wrappers based on the above `postMessage` API to achieve something similar to what you designed right now.
Depending on your use case you could also make a synchronous version of the "`postMessage` API" that returns the result with the same call, if that's simpler to you.

P.S. The video does play in VLC, but I still can't figure out what's happening.
P.S. 2: I don't have the time to run the code and debug right now, the above recommendation is based on the assumption of mixing multiple `Engine`s is the root cause.

**Reply by Zaitonn** [reply]
>Would you benefit from having multiple instances of Engine?

I use multiple instances of Engine to provide every plugin with a separate environment. Otherwise, there might be some code conflicts when running

thanks for your help very much !! I'll try it later:D

**Reply by Genteure** [reply]
[Modules](https://github.com/sebastienros/jint#using-modules) also provides some degree of separation.

**Reply by Zaitonn** [reply]
> [Modules](https://github.com/sebastienros/jint#using-modules) also provide some degree of separation.

thanks for your inspiration!!
